feat(app):  App robot settings advanced tab robot update

koji opened the pull request:
<!--
Thanks for taking the time to open a pull request! Please make sure you've read the "Opening Pull Requests" section of our Contributing Guide:

https://github.com/Opentrons/opentrons/blob/edge/CONTRIBUTING.md#opening-pull-requests

To ensure your code is reviewed quickly and thoroughly, please fill out the sections below to the best of your ability!
-->

# Overview
Added SoftwareUpdate Modal. 
This PR also includes other Advanced tab functions.

<!--
Use this section to describe your pull-request at a high level. If the PR addresses any open issues, please tag the issues here.
-->

# Changelog
- Added SoftwareUpdate Modal 
<!--
List out the changes to the code in this PR. Please try your best to categorize your changes and describe what has changed and why.

Example changelog:
- Fixed app crash when trying to calibrate an illegal pipette
- Added state to API to track pipette usage
- Updated API docs to mention only two pipettes are supported

IMPORTANT: MAKE SURE ANY BREAKING CHANGES ARE PROPERLY COMMUNICATED
-->

# Review requests
SoftwareUpdate Modal
`app/src/organisms/Devices/RobotSettings/AdvancedTab/SoftwareUpdateModal.tsx`

<!--
Describe any requests for your reviewers here.
-->

# Risk assessment
low
<!--
Carefully go over your pull request and look at the other parts of the codebase it may affect. Look for the possibility, even if you think it's small, that your change may affect some other part of the system - for instance, changing return tip behavior in protocol may also change the behavior of labware calibration.

Identify the other parts of the system your codebase may affect, so that in addition to your own review and testing, other people who may not have the system internalized as much as you can focus their attention and testing there.
-->

koji committed:
Added route for RobotSettings Advanced Tab

koji committed:
Added items to Advanced tab

koji committed:
Added items to Advanced Tab

koji committed:
Created components for each section

koji committed:
Modified style code

koji committed:
Added Download logs

koji committed:
Added robot serialNumber

koji committed:
Added Browse file system

koji committed:
Added Disable Homing

koji committed:
Added updateSettings section functionality

koji committed:
Added tests for sections that use a toggle button

koji committed:
Cleanup tests for sections that have a toggle btn

koji committed:
Added tests for RobotServer and RobotInformation

koji committed:
Added a base for Rename Robot slideout

koji committed:
Merge branch 'edge' into app_RobotSettings-Advanced-Tab

koji committed:
Added size prop to modify icon size on RobotCard

koji committed:
Added rename functionality

koji committed:
Added functionality to Factory Reset Modal

- trigger reset
- connect to a robot

koji committed:
sync with edge to open a PR
Merge branch 'edge' into app_RobotSettings-Advanced-Tab

koji committed:
Added Banner for software update

koji committed:
Merge branch 'edge' of https://github.com/Opentrons/opentrons into edge

koji committed:
sync with edge
Merge branch 'edge' into app_RobotSettings-Advanced-Tab

koji committed:
Added banner to RobotServerVersion section

koji committed:
Added the case if a robot serve version is null

koji committed:
Fixed reloading issues

koji committed:
Fixed Factory Reset modal props issue

koji committed:
Removed comments and modify connected prop

koji committed:
removed unnecessary dependency

koji committed:
Added software update modal

koji committed:
Merge branch 'edge' into app_RobotSettings-Advanced-Tab-robot-update

koji committed:
Updated the software update modal styling

b-cooper committed:
fix pipette and module card layout

b-cooper commented on app/src/atoms/Modal/index.tsx:
```diff
@@ -15,13 +15,16 @@ import {
   title?: React.ReactNode
   children?: React.ReactNode
+  icon?: IconProps
+  iconColor?: string
```
can we just use the `color` prop in IconProps, and avoid adding two new props here?

b-cooper commented on app/src/organisms/Devices/RobotSettings/AdvancedTab/AdvancedTabSlideouts/FactoryResetModal.tsx:
```diff
@@ -0,0 +1,114 @@
+            </StyledText>
+            <Flex justifyContent={JUSTIFY_FLEX_END}>
+              <SecondaryButton
+                onClick={() => {}}
```
if this is missing functionality, can you add a TODO comment or log here?

b-cooper commented on app/src/organisms/Devices/RobotSettings/AdvancedTab/AdvancedTabSlideouts/RenameRobotSlideout.tsx:
```diff
@@ -0,0 +1,96 @@
+} from '@opentrons/components'
+import { useUpdateRobotNameMutation } from '@opentrons/react-api-client'
+import { removeRobot } from '../../../../../redux/discovery'
+// import { useRobot } from '../../../hooks'
```
can this be deleted?

b-cooper commented on app/src/organisms/Devices/RobotSettings/AdvancedTab/LegacySettings.tsx:
```diff
@@ -0,0 +1,58 @@
+        <StyledText
+          as="h2"
+          css={TYPOGRAPHY.h3SemiBold}
+          //   paddingBottom={SPACING.spacing3}
```
```suggestion
```

b-cooper commented on app/src/organisms/Devices/RobotSettings/AdvancedTab/RobotInformation.tsx:
```diff
@@ -0,0 +1,74 @@
+  robot,
+}: RobotInformationProps): JSX.Element {
+  const { t } = useTranslation('device_settings')
+  // TODO: serialNumber
```
is this TODO still valid?

koji replied:
No. this branch has a function for that.

b-cooper commented on app/src/organisms/Devices/RobotSettings/AdvancedTab/RobotServerVersion.tsx:
```diff
@@ -0,0 +1,124 @@
+    'robot_server_versions_view_update'
+  )}`
+
+  // const ViewUpdateButton = (): JSX.Element => {
+  //   <Link>
+  //   </Link>
+  // }
+
+  // const SoftwareUpdateBanner = (): JSX.Element => {
+  //   return(
+
+  //   )
+  // }
+
```
```suggestion
```

b-cooper commented on app/src/organisms/Devices/RobotSettings/AdvancedTab/ShortTrashBin.tsx:
```diff
@@ -0,0 +1,55 @@
+  return (
+    <Flex alignItems={ALIGN_CENTER} justifyContent={JUSTIFY_SPACE_BETWEEN}>
+      <Box width="70%">
+        <StyledText
```
The two `StyledText`'s here should use the `as` prop instead of the `css` prop.

b-cooper commented on app/src/organisms/Devices/RobotSettings/AdvancedTab/Troubleshooting.tsx:
```diff
@@ -0,0 +1,56 @@
+      <Box width="70%">
+        <StyledText
+          as="h3"
+          css={TYPOGRAPHY.h3SemiBold}
```
```suggestion
          fontWeight={TYPOGRAPHY.fontWeightSemiBold}
```

b-cooper commented on app/src/organisms/Devices/RobotSettings/AdvancedTab/Troubleshooting.tsx:
```diff
@@ -0,0 +1,56 @@
+        >
+          {t('update_robot_software_troubleshooting')}
+        </StyledText>
+        <StyledText as="p" css={TYPOGRAPHY.pSemiBold}>
```
```suggestion
        <StyledText as="p" fontWeight={TYPOGRAPHY.fontWeightSemiBold}>
```

b-cooper commented on app/src/organisms/Devices/RobotSettings/AdvancedTab/UpdateRobotSoftware.tsx:
```diff
@@ -0,0 +1,97 @@
+      <Box width="70%">
+        <StyledText
+          as="h3"
+          css={TYPOGRAPHY.h3SemiBold}
```
```suggestion
          fontWeight={TYPOGRAPHY.fontWeightSemiBold}
```

b-cooper commented on app/src/organisms/Devices/RobotSettings/AdvancedTab/UsageSettings.tsx:
```diff
@@ -0,0 +1,57 @@
+      <Box width="70%">
+        <StyledText
+          as="h3"
+          css={TYPOGRAPHY.h3SemiBold}
```
```suggestion
          fontWeight={TYPOGRAPHY.fontWeightSemiBold}
```

b-cooper commented on app/src/organisms/Devices/RobotSettings/AdvancedTab/UsageSettings.tsx:
```diff
@@ -0,0 +1,57 @@
+        >
+          {t('usage_settings')}
+        </StyledText>
+        <StyledText as="p" css={TYPOGRAPHY.pSemiBold}>
```
```suggestion
        <StyledText as="p" fontWeight={TYPOGRAPHY.fontWeightSemiBold}>
```

b-cooper commented on app/src/organisms/Devices/RobotSettings/AdvancedTab/UseOlderAspirateBehavior.tsx:
```diff
@@ -0,0 +1,54 @@
+      <Box width="70%">
+        <StyledText
+          as="h3"
+          css={TYPOGRAPHY.h3SemiBold}
```
```suggestion
          fontWeight={TYPOGRAPHY.fontWeightSemiBold}
```

b-cooper commented on app/src/organisms/Devices/RobotSettings/AdvancedTab/UseOlderProtocol.tsx:
```diff
@@ -0,0 +1,59 @@
+      <Box width="70%">
+        <StyledText
+          as="h3"
+          css={TYPOGRAPHY.h3SemiBold}
```
```suggestion
          fontWeight={TYPOGRAPHY.fontWeightSemiBold}
```

b-cooper commented on app/src/organisms/Devices/RobotSettings/AdvancedTab/__tests__/OpenJupyterControl.test.tsx:
```diff
@@ -0,0 +1,4 @@
+// test cases
```
```suggestion
// TODO: IMMEDIATELY test cases
```

koji replied:
Added test cases

koji committed:
Merge branch 'edge' into app_RobotSettings-Advanced-Tab-robot-update

koji committed:
Fixed the problems that Brian commented

- Removed comments
- Updated StyledText props
- Added modal close function
- Added tests for OpenJupyterControl

koji committed:
Added software update modal

koji commented:
- [ ] clean up code
- [ ] fix test issue

koji committed:
Update RobotServerVersion.tsx

koji committed:
Fixed test error

- clean up code
- rename a couple of useState
- added a couple of tests to `RobotServerVersion`
- fixed errors on `RobotSettings.test.tsx` under `pages`
- Added tests to `RobotSettings.test.tsx` under `organisms`

koji committed:
Added test cases to RobotSettingAdvanced tab

koji committed:
Remove unused line

koji committed:
Fixed import statement path error

koji committed:
Fixes issues for lint-js check

koji committed:
Added basic test cases to SoftwareUpdate Modal

koji committed:
Update types.ts

koji committed:
Fixed js checks issues

koji committed:
Fixed check-js errors

koji committed:
Commented out lines for temporary patching

koji committed:
Fixed lint-js errors

koji commented:
Additional test cases and the rest of the functions will be fixed in other PRs

b-cooper committed:
abstract update robot banner and include in t1 and t2

b-cooper committed:
fix type

b-cooper committed:
swap out nested link

koji committed:
Fixed a couple things on SoftwareUpdateModal

- Added closeModal
- Updated initial useState for showSoftwareUpdateModal

mcous commented on discovery-client/src/store/__tests__/actions.test.ts:
```diff
@@ -56,12 +56,14 @@ describe('discovery client action creators', () => {
     const health = {
       name: 'opentrons-dev',
       api_version: '1.2.3',
+      serial_number: '12345',
```
Is this correct? IIRC regular `/health` does not include serial number, only `/server/health`

mcous commented on api-client/src/server/updateRobotName.ts:
```diff
@@ -0,0 +1,17 @@
+  config: HostConfig,
+  newName: string
+): ResponsePromise<UpdatedRobotName> {
+  const newRobotName = JSON.stringify({ name: newName })
```
Unnecessary `JSON.stringify`

b-cooper committed:
mock out update banner in robot card and overview

koji committed:
Fixed what Mike commented

koji committed:
Fixed the check js issue

b-cooper committed:
remove robot card update banner temporarily because of click propagation bug

b-cooper committed:
remove unused import

b-cooper reviewed (approved):
🚢

mcous commented on api-client/src/health/types.ts:
```diff
@@ -1,6 +1,7 @@
 export interface Health {
   name: string
   api_version: string
+  serial_number: string
```
This should be removed

mcous commented on api-client/src/server/updateRobotName.ts:
```diff
@@ -7,11 +7,10 @@ export function updateRobotName(
     POST,
     '/server/name',
-    newRobotName,
+    newName,
```
Oops, this wasn't what I meant! The `JSON.stringify` was unnecessary, but the object was absolutely required. This is probably the cause of #10036.

It should have been:

```ts
return request<UpdatedRobotName, { name: string }>(
  POST,
  '/server/name',
  { name: newName },
  config
)
```

koji replied:
Thank you for the suggestion.

mcous reviewed (commented):
I believe this is the cause of #10036

koji commented:
Thank you, @mcous. 
I updated my working branch to deliver the rename function
